[Runtime] Avoid function pointer indirection in refcounting functions.

mikeash opened the pull request:
Functions like swift_retain call through a function pointer so that Instruments can interpose. This slows down the common case where there is no interposition. Instead, initialize the function pointers to NULL and call through directly to the real implementation when it's NULL. The compiler is smart enough to inline this call and the result is a single conditional branch rather than a function pointer call.

rdar://problem/18307425

mikeash committed:
[Runtime] Avoid function pointer indirection in refcounting functions.

Functions like swift_retain call through a function pointer so that Instruments can interpose. This slows down the common case where there is no interposition. Instead, initialize the function pointers to NULL and call through directly to the real implementation when it's NULL. The compiler is smart enough to inline this call and the result is a single conditional branch rather than a function pointer call.

rdar://problem/18307425

mikeash commented:
@swift-ci please benchmark

compnerd commented on stdlib/public/runtime/HeapObject.cpp:
```diff
@@ -98,7 +92,18 @@ static HeapObject *_swift_allocObject_(HeapMetadata const *metadata,
+                                     size_t requiredAlignmentMask) {
+  if (SWIFT_UNLIKELY(_swift_allocObject))
+    return _swift_allocObject(metadata, requiredSize, requiredAlignmentMask);
+  else
```
The `else` is unnecessary, and similar throughout.

mikeash replied:
Thanks, updated.

rjmccall commented:
Very nice!

mikeash committed:
[Runtime] Initialize function pointers with the original values, check against those values to call directly.

Instruments relies on the old values being there so it can call the original implementation. This has very slightly worse codegen but the impact should be minimal.

mikeash commented:
Turns out that Instruments doesn't work when the function pointers start out as `NULL`. Trying a different approach now where we initialize them as before, but then we check for that value and make the call directly. The compiler still inlines the implementation in this case. One additional `leaq` instruction is generated to get the address to compare with, which shouldn't hurt too much.

mikeash commented:
@swift-ci please benchmark
